### PR TITLE
Average extinction curve

### PR DESCRIPTION
**Pull request by mdecleir**
I added functionality to calculate and plot average extinction curves.
Also the calculation of the uncertainties on the average curve have been revised (see discussion below).

### CODE REVIEW

**Comment by mdecleir** [comment]
Following the discussion of this morning, I updated the calculation of the average extinction curve to be more general. I have one remaining question about the uncertainty on the average. It is currently calculated as:
sqrt(sum of squared uncertainties/number of curves) / number of curves. But, what do we actually want to calculate here? The standard deviation around the mean, or the standard error of the mean, or something else. I was reading this: https://www.investopedia.com/ask/answers/042415/what-difference-between-standard-error-means-and-standard-deviation.asp, where they explain the difference between the standard deviation and the standard error, but they use the square of the difference between every value and the mean, and not the square of the uncertainties...

**Comment by karllark** [comment]
What do you think we need for the average?

**Comment by karllark** [comment]
After further thought, I'm leaning towards the empirical sample based standard deviation of the mean (e.g., based on the differences between each curve and the mean instead of the propagated uncertainties).  But would very much like to hear your thoughts on this.

**Comment by mdecleir** [comment]
So, as it has been ages since I got statistics at university, I spent some time searching for resources, and I obviously opened a can of worms... However, with the help of my course notes on my parents' atic, and wikipedia, I think I now more or less understand what the differences are, and what to use in which situation. Here are my findings:

1) Propagating the uncertainties (i.e., adding the individual uncertainties quadratically as is done in the code) is used to measure the uncertainty on a function of DIFFERENT variables, e.g. x+y. This is thus NOT what we need, since we are adding different measurements of the SAME variable.

2) When measuring an average (or mean), one usually also measures the _**standard deviation**_. This is the variation of the individual measurements around the mean, calculated as:
Screen Shot 2020-12-17 at 1 51 28 PM
OR
Screen Shot 2020-12-17 at 1 51 35 PM
Whether you use N or N-1 depends on whether you want the SD of the sample or of the entire population. From https://www.graphpad.com/support/faqid/1383/:

_The n-1 equation is used in the common situation where you are analyzing a sample of data and wish to make more general conclusions. The SD computed this way (with n-1 in the denominator) is your best guess for the value of the SD in the overall population. If you simply want to quantify the variation in a particular set of data, and don't plan to extrapolate to make wider conclusions, then you can compute the SD using n in the denominator. The resulting SD is the SD of those particular values._

I would argue we want N in the denominator since we are interested in the variation of our sample around the mean of the sample (not around the mean of the population, which is unknown anyway).

3) Finally, there is a difference between the **_standard deviation_** and the **_standard error_**: From https://en.wikipedia.org/wiki/Standard_error:

_the standard error of the sample mean is an estimate of how far the sample mean is likely to be from the population mean, whereas the standard deviation of the sample is the degree to which individuals within the sample differ from the sample mean._
The standard error of the mean is calculated as:

Screen Shot 2020-12-17 at 2 08 29 PM, with sigma the standard deviation of the population, which is unknown, but can be estimated by _s_ from 2).

So, I think what we need is the standard deviation of the sample.


In conclusion, I believe we want to calculate this:
Screen Shot 2020-12-17 at 1 51 28 PM

I hereby conclude this lesson in statistics...

**Comment by karllark** [comment]
That was a good lesson in statistics.  :-)

I'm not sure I agree that we want the standard deviation of the sample.  The goal of averaging a set of extinction curves is to get the average behavior.  If I was going to use this average in forward modeling or removing extinction from my observations, I would want to know the uncertainty of using it and this would be the standard deviation of the mean as it is a mean I'm using.  Knowing the standard deviation of the sample may be useful, but mainly to ask if another measured extinction curve is consistent with the mean, right?

**Comment by karllark** [comment]
Or maybe both need to be calculated and reported from this function?  Then the user can pick the one that is key for their analysis?

**Comment by mdecleir** [comment]
> I'm not sure I agree that we want the standard deviation of the sample. The goal of averaging a set of extinction curves is to get the average behavior. If I was going to use this average in forward modeling or removing extinction from my observations, I would want to know the uncertainty of using it and this would be the standard deviation of the mean as it is a mean I'm using. Knowing the standard deviation of the sample may be useful, but mainly to ask if another measured extinction curve is consistent with the mean, right?

Ok yes, if you look at it in this way, you are right. In that case, what we want is the standard error on the mean, using an estimate for the standard deviation with N-1 in the denominator, because we want to generalize to the entire Milky Way.

**Comment by mdecleir** [comment]
> Or maybe both need to be calculated and reported from this function? Then the user can pick the one that is key for their analysis?

Good idea, I will calculate both and name them differently. The standard deviation can be seen as the spread of the curves around the mean, i.e. how good is our sample?

**Comment by karllark** [comment]
Sounds great.

**Comment by mdecleir** [comment]
I guess the final question is where to "store" these different uncertainties. The average extinction is just an object of ExtData, which only has 1 uncs keyword, and having more than 1 keyword for uncertainties doesn't make sense for individual ExtData objects, but only for the average extinction.

**Comment by karllark** [comment]
You can just add a new variable is the extdata class.  Maybe make it sampleunc or something like that.  Then the existence of this variable can be check when saving and added to the output if it does exist.  Same with reading.

File: measure_extinction/extdata.py [hunk 1]
```diff
@@ -97,81 +98,77 @@ def _get_column_val(column):
...
+                aveext.stds["BAND"].append(np.nanstd(bexts[name], ddof=1))
+
+                # calculation of the standard error of the average (the standard error of the sample mean is an estimate of how far the sample mean is likely to be from the population mean)
+                aveext.uncs["BAND"].append(stats.sem(bexts[name]))
+
+        else:
+            aveext.exts[src] = np.nanmean(exts, axis=0)
+            aveext.npts[src] = np.sum(~np.isnan(exts), axis=0)
+            aveext.stds[src] = np.nanstd(exts, axis=0, ddof=1)
+            aveext.uncs[src] = stats.sem(exts, axis=0, nan_policy="omit")
```
Comment (karllark) [inline measure_extinction/extdata.py:171]: Why not just calculate the uncs from stds directly?  Shouldn't uncs = stds/sqrt(npts)?

File: measure_extinction/extdata.py [hunk 2]
```diff
@@ -201,9 +198,11 @@ class ExtData:
     waves : dict of key:wavelengths
         key is BAND, IUE, IRS, etc.
 
-    ext : dict of key:E(lambda-v) measurements
+    ext : dict of key:E(lambda-v) or A(lambda)/A(V) measurements
 
-    uncs : dict of key:E(lambda-v) measurement uncertainties
+    uncs : dict of key:E(lambda-v) or A(lambda)/A(V) measurement uncertainties
```
Comment (karllark) [inline measure_extinction/extdata.py:203]: Instead of "V",this should be "x" as the relative band is now set by type_rel_band.

File: measure_extinction/extdata.py [hunk 3]
```diff
@@ -201,9 +198,11 @@ class ExtData:
...
+
+    stds : dict of key:A(lambda)/A(V) standard deviations (this is the spread of the sample around the population mean)
```
Comment (karllark) [inline measure_extinction/extdata.py:205]: Say something about this is only defined if the curve is an average of a set of curves?

File: measure_extinction/extdata.py [hunk 4]
```diff
@@ -235,6 +234,7 @@ def __init__(self, filename=None):
         self.waves = {}
         self.exts = {}
         self.uncs = {}
+        self.stds = {}
```
Comment (karllark) [inline measure_extinction/extdata.py:237]: I'm not sure that this should be defined by default?  Won't this mean the write function will always try and write this column even if it hasn't been calculated?

**Review by karllark — CHANGES_REQUESTED** [review]
A few comments.

File: measure_extinction/extdata.py (hunk 1 above)
Comment (mdecleir) [reply measure_extinction/extdata.py:171]: That is also possible, yes. I used the built-in function because it exists... Do you want me to change it?

File: measure_extinction/extdata.py (hunk 4 above)
Comment (mdecleir) [reply measure_extinction/extdata.py:237]: Are you referring to the save function? The save function has
`if curname in self.stds.keys()` and should thus only write this column if the stds has been calculated for that part of the curve...

Comment (mdecleir) [reply measure_extinction/extdata.py:237]: If I do not define `self.stds` as an empty library, I get an AttributeError: 'ExtData' object has no attribute 'stds' when trying to add the standard deviation...

Comment (karllark) [reply measure_extinction/extdata.py:237]: But isn't the stds part of the keys() even if it is an empty dictonary?
When you calculate stds in the averaging process, define it there before you use it.  You can add to the class variables anywhere you want.

File: measure_extinction/extdata.py (hunk 1 above)
Comment (mdecleir) [reply measure_extinction/extdata.py:171]: I changed it.

File: measure_extinction/extdata.py (hunk 4 above)
Comment (mdecleir) [reply measure_extinction/extdata.py:237]: The keys() are e.g. IUE, BAND, etc... If I try to define stds in the averaging process, I get an error:
`aveext.stds = {}`
AttributeError: 'ExtData' object has no attribute 'stds'

Comment (karllark) [reply measure_extinction/extdata.py:237]: Good point.  I retract my comment.

**Comment by mdecleir** [comment]
I don't understand why the dev test fails... It works on my laptop locally... Can you restart the testing?

**Review by karllark — APPROVED** [review]
Looks great.
